### PR TITLE
middleware: Tests for Nop* and Chain middleware

### PR DESCRIPTION
**Pull request by abhinav**
The chain tests also ensure that we're okay to call the outbound/handler
multiple times from a middleware.

### CODE REVIEW

**Comment by abhinav** [comment]
@yarpc/golang

File: internal/filter/chain_test.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+	"github.com/yarpc/yarpc-go/encoding/raw"
+	"github.com/yarpc/yarpc-go/transport"
+	"github.com/yarpc/yarpc-go/transport/transporttest"
+
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/assert"
+	"golang.org/x/net/context"
+)
+
+var retryFilter = transport.FilterFunc(
```
Comment (prashantv) [inline internal/filter/chain_test.go]: optional: might be cleaner to write:
```go
var retryFilter transport.FilterFunc = func(.....)
```

Comment (abhinav) [inline internal/filter/chain_test.go]: oh, cool. TIL

File: internal/filter/chain_test.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+	ctx context.Context, req *transport.Request, o transport.Outbound) (*transport.Response, error) {
+	c.Count++
+	return o.Call(ctx, req)
+}
+
+func TestChain(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	before := &countFilter{}
```
Comment (prashantv) [inline internal/filter/chain_test.go]: nit: move these to before you call `ApplyFilter`

File: internal/filter/chain_test.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+	o.EXPECT().Call(ctx, req).After(
+		o.EXPECT().Call(ctx, req).Return(nil, errors.New("great sadness")),
+	).Return(res, nil)
+
+	gotRes, err := transport.ApplyFilter(
+		o, Chain(before, retryFilter, after)).Call(ctx, req)
+
+	assert.NoError(t, err, "expected success")
+	assert.Equal(t, 1, before.Count, "expected outer filter to be called once")
+	assert.Equal(t, 2, after.Count, "expected outer filter to be called once")
```
Comment (prashantv) [inline internal/filter/chain_test.go]: error message is wrong

File: internal/filter/chain_test.go (hunk 2 above)
Comment (abhinav) [inline internal/filter/chain_test.go]: 👍

File: internal/filter/chain_test.go (hunk 3 above)
Comment (abhinav) [inline internal/filter/chain_test.go]: oh derp

**Comment by prashantv** [comment]
lgtm

**Pull request merged**